### PR TITLE
Rename TemporalMemory extra to externalPredictiveInputs

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
There was confusion about what this feature does so I renamed it
for clarity.

See also issue #512.

### CODE REVIEW

File: src/htm/algorithms/TemporalMemory.hpp [hunk 1]
```diff
@@ -113,11 +113,11 @@ class TemporalMemory : public Serializable
    * duplicates. Disable this for a small speed boost.
    * DEPRECATED: The SDR class now enforces these properties.
    *
-   * @param extra
+   * @param externalPredictiveInputs
    * Number of external predictive inputs.  These inputs are used in addition to
    * the cells which are part of this TemporalMemory.  The TemporalMemory
    * requires all external inputs be identified by an index in the
-   * range [0, extra).
```
Comment (breznak) [inline src/htm/algorithms/TemporalMemory.hpp]: maybe add? : 
a) If specified, `compute()` will expect additional fied with values in range [0, extraPred) for each iteration. 
b) These values are not related to the current TM, they represent input from a different region.

Comment (ctrl-z-9000-times) [reply src/htm/algorithms/TemporalMemory.hpp]: done

File: src/htm/algorithms/TemporalMemory.hpp [hunk 2]
```diff
@@ -99,6 +99,11 @@ class TemporalMemory : public Serializable
    * @param predictedSegmentDecrement
    * Amount by which segments are punished for incorrect predictions.
    *
+   * Note: A good value is just a bit larger than
+   * (the column-level sparsity * permanenceIncrement). So, if column-level
```
Comment (breznak) [inline src/htm/algorithms/TemporalMemory.hpp]: OT: #279 it would be nice to reduce the number of parameters for the algorithms, especially in cases like this where a good-enough value can be derived from other settings. Do you think this would be a reasonable candidate to hard-code? 
Pros: 
- less space for user to mess up
- faster parm optimizations
Cons: 
- removes a bit of configurability

Comment (ctrl-z-9000-times) [reply src/htm/algorithms/TemporalMemory.hpp]: I agree, but here are two more potential issues:
1) Is this advice always good?  I think it would need further investigation before we remove the parameter in favor of this heuristic.
2) We would have to modify every instance of the TM to omit this parameter.  It would be a lot of work.
